pkg/apply/applier_apply_options.go

```diff
@@ -0,0 +1,141 @@
+	// Encapsulates the fields for server-side apply.
+	serverSideOptions common.ServerSideOptions
+
+	// reconcileTimeout defines whether the applier should wait
+	// until all applied resources have been reconciled, and if so,
+	// how long to wait.
+	reconcileTimeout time.Duration
+
+	// pollInterval defines how often we should poll for the status
+	// of resources.
+	pollInterval time.Duration
+
+	// emitStatusEvents defines whether status events should be
+	// emitted on the eventChannel to the caller.
+	emitStatusEvents bool
+
+	// prune defines whether pruning of previously applied
+	// objects should happen after apply.
+	prune bool
+
+	// dryRunStrategy defines whether changes should actually be performed,
+	// or if it is just talk and no action.
+	dryRunStrategy common.DryRunStrategy
+
+	// prunePropagationPolicy defines the deletion propagation policy
+	// that should be used for pruning. If this is not provided, the
+	// default is to use the Background policy.
+	prunePropagationPolicy metav1.DeletionPropagation
+
+	// pruneTimeout defines whether we should wait for all resources
+	// to be fully deleted after pruning, and if so, how long we should
+	// wait.
+	pruneTimeout time.Duration
+
+	// inventoryPolicy defines the inventory policy of apply.
+	inventoryPolicy inventory.InventoryPolicy
+
+	// eventListeners is a list of event listeners that are called in specified order when an event happens.
+	// Listeners are never called concurrently with each other.
+	eventListeners []func(event.Event)
+}
+
+func NewOptions() *Options {
+	return &Options{
+		pollInterval:           defaultPollInterval,
+		prune:                  true,
+		prunePropagationPolicy: metav1.DeletePropagationBackground,
+	}
+}
+
+func (o *Options) ServerSideOptions(serverSideOptions common.ServerSideOptions) *Options {
+	o.serverSideOptions = serverSideOptions
+	return o
+}
+
+func (o *Options) ReconcileTimeout(reconcileTimeout time.Duration) *Options {
+	o.reconcileTimeout = reconcileTimeout
+	return o
+}
+
+func (o *Options) PollInterval(pollInterval time.Duration) *Options {
+	o.pollInterval = pollInterval
+	return o
+}
+
+func (o *Options) EmitStatusEvents(emitStatusEvents bool) *Options {
+	o.emitStatusEvents = emitStatusEvents
+	return o
+}
+
+func (o *Options) Prune(prune bool) *Options {
+	o.prune = prune
+	return o
+}
+
+func (o *Options) DryRunStrategy(dryRunStrategy common.DryRunStrategy) *Options {
+	o.dryRunStrategy = dryRunStrategy
+	return o
+}
+
+func (o *Options) PrunePropagationPolicy(prunePropagationPolicy metav1.DeletionPropagation) *Options {
+	o.prunePropagationPolicy = prunePropagationPolicy
+	return o
+}
+
+func (o *Options) PruneTimeout(pruneTimeout time.Duration) *Options {
+	o.pruneTimeout = pruneTimeout
+	return o
+}
+
+func (o *Options) InventoryPolicy(inventoryPolicy inventory.InventoryPolicy) *Options {
+	o.inventoryPolicy = inventoryPolicy
+	return o
+}
+
+func (o *Options) EventListener(eventListener ...func(event.Event)) *Options {
```

Comment:
These event channel changes should probably go in a separate PR. They seem enabled by but not required for the Builder pattern. If you leave the applier.Run the way it was, it should significantly simplify this PR.

I'm not sure how I feel about these EventChannel changes, but it doesn't need to block the other changes.